### ISSUE TITLE
feat: support sourcehut

### ISSUE THREAD

**Issue by cafkafk** [cafkafk/rime#14]
I'm unsure if we can, we probably will not be able to get releases. But! We might be able to default to an endpoint that does a bit of thinking and finds the latest semver tag.

Of course, if we make such logic, I'd like to have it be less endpoing specific, because it could be extremely useful for e.g. #10.

I already created a testing repo that might be used in integration tests https://git.sr.ht/~cafkafk/hello

**Comment by cafkafk** [comment]
I found that, if tagged, we can get a tarball from sourcehut[1].

Example:

```sh
nix run https://git.sr.ht/~cafkafk/hello/archive/v0.0.1.tar.gz
```

[1]: https://lists.sr.ht/~sircmpwn/sr.ht-discuss/%3Cf7af3c4e-ee62-95dd-a8fd-83fbd6a88ca7%40ecmelberk.com%3E

**Comment by algernon** [comment]
That message mentions that tarballs exist for all git refs too, they're just not shown in the web UI.

```
❯ curl -I 'https://git.sr.ht/~cafkafk/hello/archive/main.tar.gz'
HTTP/2 200
[...snip...]
```

**Comment by cafkafk** [comment]
that's easy /version/branch/refs endpoints

**Comment by algernon** [comment]
> I'm unsure if we can, we probably will not be able to get releases. But! We might be able to default to an endpoint that does a bit of thinking and finds the latest semver tag.

I did a bit of digging, and so far my finding is that we can't do that. While sourcehut has an [api](https://man.sr.ht/git.sr.ht/api.md), most (all?) of it is locked behind an access token, so we'd have to resort to html scraping, and... I don't think we should do that.

It might be possible to supply an access token to Rime in its configuration, perhaps, optionally?

**Comment by cafkafk** [comment]
We might at some point need to support access tokens... maybe it is inevitable... but if we can avoid it we really should. 

I did try `curl -L https://git.sr.ht/api/~cafkafk/repos/hello/refs` and it explicitly says:

```
 {"errors": [{"reason": "No authorization supplied (expected an OAuth token)"}]}⏎ 
 ```
 
 So that's a bummer. I tried the endpoint with a personal access token, and that did seem to work at least, and gives anything we might need. The best option here is to use a personal access token right? https://man.sr.ht/meta.sr.ht/oauth-api.md#personal-access-tokens
 
 > It might be possible to supply an access token to Rime in its configuration, perhaps, optionally?
 
 I mean let's go that route, I think we should aim to implement any config option as being optional, and I suppose we will just add some error messages to routes that require an access token being present if it's None.
